src/experimental/tools/download2go/utils.ts

```diff
@@ -0,0 +1,255 @@
+    );
+  }
+
+  if (activeDownloads[manifest.contentID] != null) {
+    throw new ValidationArgsError("The content is already downloading");
+  }
+
+  if (manifest.progress.percentage === 100) {
+    throw new ValidationArgsError(
+      "You can't resume a content that is already fully downloaded"
+    );
+  }
+}
+
+export function checkForPauseAMovie(contentID: string) {
+  if (contentID == null || contentID === "") {
+    throw new ValidationArgsError(
+      "A valid contentID is mandatory when pausing"
+    );
+  }
+}
+
+// DRM Capabilities:
+
+// Key Systems
+const CENC_KEY_SYSTEMS = [
+  "com.widevine.alpha",
+  "com.microsoft.playready.software",
+  "com.apple.fps.1_0",
+  "com.chromecast.playready",
+  "com.youtube.playready",
+];
+
+// Robustness ONLY FOR WIDEVINE
+const WIDEVINE_ROBUSTNESSES = [
+  "HW_SECURE_ALL",
+  "HW_SECURE_DECODE",
+  "HW_SECURE_CRYPTO",
+  "SW_SECURE_DECODE",
+  "SW_SECURE_CRYPTO",
+];
+
+/**
+ * Construct the necessary configuration for getCompatibleDRMConfigurations() Prober tool
+ *
+ * @returns {Array.<{type: String, configuration: Object<MediaKeySystemConfiguration>}>}
+ */
+export function constructConfig(): Array<{
+  type: string;
+  configuration: IMediaKeySystemConfiguration;
+}> {
+  return CENC_KEY_SYSTEMS.map(keySystem => ({
+    type: keySystem,
+    configuration: getKeySystemConfigurations(keySystem),
+  }));
+}
+
+/**
+ * @param {string} keySystem
+ * @returns {MediaKeySystemConfiguration[]}
+ */
+export function getKeySystemConfigurations(
+  keySystem: string
+): IMediaKeySystemConfiguration {
+  const videoCapabilities: MediaKeySystemMediaCapability[] = [];
+  const audioCapabilities: MediaKeySystemMediaCapability[] = [];
+  const robustnesses =
+    keySystem === "com.widevine.alpha"
+      ? WIDEVINE_ROBUSTNESSES
+      : [undefined, undefined, undefined, undefined];
+
+  robustnesses.forEach((robustness: string | undefined) => {
+    videoCapabilities.push({
+      contentType: "video/mp4;codecs='avc1.4d401e'", // standard mp4 codec
+      robustness,
+    });
+    videoCapabilities.push({
+      contentType: "video/mp4;codecs='avc1.42e01e'",
+      robustness,
+    });
+    videoCapabilities.push({
+      contentType: "video/webm;codecs='vp8'",
+      robustness,
+    });
+    audioCapabilities.push({
+      contentType: "audio/mp4;codecs='mp4a.40.2'", // standard mp4 codec
+      robustness,
+    });
+  });
+
+  return {
+    initDataTypes: ["cenc"],
+    videoCapabilities,
+    audioCapabilities,
+    persistentState: "required",
+    sessionTypes: ["persistent-license"],
+  };
+}
+
+function isFairplayDrmSupported(): boolean {
+  const MK = (window as any).WebKitMediaKeys as {
+    isTypeSupported: (drm: string, applicationType: string) => boolean;
+  };
+  const drm = "com.apple.fps.1_0";
+  return (
+    MK !== undefined &&
+    MK.isTypeSupported !== undefined &&
+    MK.isTypeSupported(drm, "video/mp4")
+  );
+}
+
+function isSafari() {
```

Comment:
Most of the time, the **userAgent** is not very reliable to detect an environment because for some reason it could be changed. Instead, the one, we are using in the `OnePlayer` seems more robust since it checks the presence of additional API only present in **Safari**.

I could add/replace this `function` with the one you are currently using.

We are using this one:
```js
export const isSafari = () =>
  Object.prototype.toString.call(window.HTMLElement).includes('Constructor') ||
  (!window.safari || safari.pushNotification).toString() ===
    '[object SafariRemoteNotification]';
```

Tell me how to proceed if we go further :)